src/classes/EP_TaskDependencyCheck_TDTM.cls

```diff
@@ -0,0 +1,170 @@
+                FROM Engagement_Plan_Tasks__r)
+            FROM Engagement_Plan_Template__c WHERE Id IN :templateIds]);
+        
+        // Loop through the lists of tasks check for dependency recursion and ensure template matches the dependent tasks
+        for (Engagement_Plan_Task__c newEPTask : listToValidate) {
+            Map<Id, Engagement_Plan_Task__c> mapExistingEPTasks =
+                new Map<Id, Engagement_Plan_Task__c>(
+                    mapIdToTemplateWithTasks.get(newEPTask.Engagement_Plan_Template__c).Engagement_Plan_Tasks__r);
+            if (newEPTask.Dependent_Task__c != null) {
+                // If the dependent task does not belong to the same template add an error
+                if (!mapExistingEPTasks.containsKey(newEPTask.Dependent_Task__c)) {
+                    newEPTask.addError(Label.EPTaskTemplate);
+                } else {
+                    // Traverse through the dependency chain to verify the dependency only occurs once in the chain
+                    Set<Id> taskIds = new Set<Id>{newEPTask.Id};
+                    Engagement_Plan_Task__c dependentTask = mapExistingEPTasks.get(newEPTask.Dependent_Task__c);
+                    Integer count = 0;
+                    while (dependentTask.Dependent_Task__c != null
+                            && mapExistingEPTasks.containsKey(dependentTask.Dependent_Task__c)
+                            && count < maxCount) {
+                        if (taskIds.contains(dependentTask.Dependent_Task__c)) {
+                            newEPTask.addError(Label.EPTaskDependency);
+                            break;
+                        }
+                        taskIds.add(dependentTask.Dependent_Task__c);
+                        dependentTask = mapExistingEPTasks.get(dependentTask.Dependent_Task__c);
+                        count++;
+                    }
+                }
+            }
+        }
+    }
+    /*******************************************************************************************************
+    * @description For the Tasks that are updated and have the EP Template changed,
+    * query for any child tasks to verify the templates are the same
+    * @param listToValidate the list of EP Tasks that require child dependency validation. 
+    * @return void.  
+    ********************************************************************************************************/
+    private static void validateChildDependencies(List<Engagement_Plan_Task__c> listToValidate) {
+        // Engagement Plan Task mapped to tasks that are dependent on it to check for template mismatch
+        Map<Id, Engagement_Plan_Task__c> mapIdToTaskWithChildren = new Map<Id, Engagement_Plan_Task__c>(
+            [SELECT Id, 
+                (SELECT Id, Dependent_Task__c, Engagement_Plan_Template__c
+                FROM Action_Plan_Tasks__r)
```

Comment:
Sorry, this is a pain, but would you mind changing the child relationship name to start with Engagement_ instead of Action_?